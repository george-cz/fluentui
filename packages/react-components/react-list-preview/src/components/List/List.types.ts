import * as React from 'react';
import type { ComponentProps, ComponentState, Slot, SelectionMode, SelectionItemId } from '@fluentui/react-utilities';
import { ListSelectionState } from '../../hooks/types';

export type ListLayout = 'horizontal' | 'vertical' | 'grid';

export type ListSlots = {
  root: NonNullable<Slot<'ul', 'div' | 'ol'>>;
};

/**
 * List Props
 */
export type ListProps = ComponentProps<ListSlots> & {
  layout?: ListLayout;
  focusableItems?: boolean;
  selectable?: boolean;
  selectionMode?: SelectionMode;
  selectedItems?: SelectionItemId[];
  defaultSelectedItems?: SelectionItemId[];
  onSelectionChange?: (event: React.SyntheticEvent, data: { selectedItems: SelectionItemId[] }) => void;
};

export type ListContextValue = {
  focusableItems: boolean;
  items: Array<{ id: string | number }>;
  selection?: ListSelectionState;
  registerItem?: (id: string | number, ref: React.RefObject<HTMLElement>) => void;
  deregisterItem?: (id: string | number, ref: React.RefObject<HTMLElement>) => void;
  as?: 'div' | 'ol' | 'ul';
};

export type ListContextValues = {
  listContext: ListContextValue;
};

/**
 * State used in rendering List
 */
export type ListState = ComponentState<ListSlots> & Required<Pick<ListProps, 'layout'>> & ListContextValue;
