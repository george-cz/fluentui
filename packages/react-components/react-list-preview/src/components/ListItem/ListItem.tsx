import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useListItem_unstable } from './useListItem';
import { renderListItem_unstable } from './renderListItem';
import { useListItemStyles_unstable } from './useListItemStyles.styles';
import type { ListItemProps } from './ListItem.types';

/**
 * ListItem component - TODO: add more docs
 */
export const ListItem: ForwardRefComponent<ListItemProps> = React.forwardRef<HTMLElement>((props, ref) => {
  const state = useListItem_unstable(props, ref);

  useListItemStyles_unstable(state);
  useCustomStyleHook_unstable('useListItemStyles_unstable')(state);
  return renderListItem_unstable(state);
});

ListItem.displayName = 'ListItem';
