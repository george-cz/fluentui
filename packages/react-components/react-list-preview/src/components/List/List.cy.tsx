import * as React from 'react';
import { mount as mountBase } from '@cypress/react';
import { FluentProvider } from '@fluentui/react-provider';
import { teamsLightTheme } from '@fluentui/react-theme';

import { List } from './List';
import { ListItem } from '../ListItem';

const mount = (element: JSX.Element) => {
  mountBase(<FluentProvider theme={teamsLightTheme}>{element}</FluentProvider>);
};

/**
 * Validates focus movement based on the sequence of keybaord commands
 * Use focused:<data-test> to validate the focused element after a given step
 * @param sequence - Array of commands to execute
 * @example
 * testSequence([
 *  'focused:list-item-1',
 *  'DownArrow',
 *  'focused:list-item-2',
 * ])
 */
const testSequence = (sequence: Array<string>) => {
  cy.get('li:first-of-type').focus();
  for (const command of sequence) {
    if (command.startsWith('focused:')) {
      const tid = command.split(':')[1];
      cy.focused().should('have.attr', 'data-test', tid);
    } else {
      cy.focused().type(`{${command}}`);
    }
  }
};

const mountSimpleList = () => {
  mount(
    <List>
      <ListItem tabIndex={0} data-test="list-item-1">
        List Item 1
      </ListItem>
      <ListItem tabIndex={0} data-test="list-item-2">
        List Item 2
      </ListItem>
      <ListItem tabIndex={0} data-test="list-item-3">
        List Item 3
      </ListItem>
    </List>,
  );
};

const mountListWithSecondaryActions = () => {
  mount(
    <List>
      <ListItem tabIndex={0} data-test="list-item-1">
        List Item 1<button data-test="list-item-1-button-1">1:button1</button>
        <button data-test="list-item-1-button-2">1:button2</button>
      </ListItem>
      <ListItem tabIndex={0} data-test="list-item-2">
        List Item 2<button data-test="list-item-2-button-1">2:button1</button>
        <button data-test="button-2-2">2:button2</button>
      </ListItem>
      <ListItem tabIndex={0} data-test="list-item-3">
        List Item 3<button data-test="button-3-1">3:button1</button>
        <button data-test="button-3-2">3:button2</button>
      </ListItem>
    </List>,
  );
};

describe('List', () => {
  describe('keyboard navigation', () => {
    describe('Simple list with a single action', () => {
      it('Up/Down arrow keys work', () => {
        mountSimpleList();
        testSequence([
          'focused:list-item-1',
          'DownArrow',
          'focused:list-item-2',
          'DownArrow',
          'focused:list-item-3',
          'DownArrow',
          'focused:list-item-3',
          'UpArrow',
          'focused:list-item-2',
          'UpArrow',
          'focused:list-item-1',
          'UpArrow',
          'focused:list-item-1',
        ]);
      });

      it('Home/End arrow keys work', () => {
        mountSimpleList();
        testSequence(['focused:list-item-1', 'End', 'focused:list-item-3', 'Home', 'focused:list-item-1']);
      });

      it('PgUp/PgDown arrow keys work', () => {
        mountSimpleList();
        testSequence(['focused:list-item-1', 'PageDown', 'focused:list-item-3', 'PageUp', 'focused:list-item-1']);
      });
    });
    describe('List with multiple actions', () => {
      it('Up/Down arrows work', () => {
        mountListWithSecondaryActions();
        testSequence([
          'focused:list-item-1',
          'DownArrow',
          'focused:list-item-2',
          'DownArrow',
          'focused:list-item-3',
          'DownArrow',
          'focused:list-item-3',
          'UpArrow',
          'focused:list-item-2',
          'UpArrow',
          'focused:list-item-1',
          'UpArrow',
          'focused:list-item-1',
        ]);
      });

      it('Home/End arrow keys work', () => {
        mountListWithSecondaryActions();
        testSequence(['focused:list-item-1', 'End', 'focused:list-item-3', 'Home', 'focused:list-item-1']);
      });

      it('PgUp/PgDown arrow keys work', () => {
        mountListWithSecondaryActions();
        testSequence(['focused:list-item-1', 'PageDown', 'focused:list-item-3', 'PageUp', 'focused:list-item-1']);
      });

      it('Left/Right arrow key moves focus horizontally in the list item', () => {
        mountListWithSecondaryActions();
        testSequence([
          'focused:list-item-1',
          'RightArrow',
          'focused:list-item-1-button-1',
          'RightArrow',
          'focused:list-item-1-button-2',
          'RightArrow',
          'focused:list-item-1-button-2',
          'LeftArrow',
          'focused:list-item-1-button-1',
          'LeftArrow',
          'focused:list-item-1',
        ]);
      });

      it('Escape moves out of the secondary and focuses on the same row', () => {
        mountListWithSecondaryActions();
        testSequence([
          'focused:list-item-1',
          'RightArrow',
          'focused:list-item-1-button-1',
          'RightArrow',
          'focused:list-item-1-button-2',
          'Esc',
          'focused:list-item-1',
        ]);
      });

      it('Arrow up/down on the secondary action focuses the item above/below', () => {
        mountListWithSecondaryActions();
        testSequence([
          'focused:list-item-1',
          'DownArrow',
          'RightArrow',
          'focused:list-item-2-button-1',
          'UpArrow',
          'focused:list-item-1',
          'DownArrow',
          'RightArrow',
          'focused:list-item-2-button-1',
          'DownArrow',
          'focused:list-item-3',
        ]);
      });

      it('Keys like Enter and Space are ignored', () => {
        mountListWithSecondaryActions();
        testSequence(['RightArrow', 'focused:list-item-1-button-1', 'Enter', ' ', 'focused:list-item-1-button-1']);
      });
    });
  });
});
