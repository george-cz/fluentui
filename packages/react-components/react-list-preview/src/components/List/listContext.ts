import * as React from 'react';
import { ListContextValue } from './List.types';

const listContext = React.createContext<ListContextValue | undefined>(undefined);

export const listContextDefaultValue: ListContextValue = {
  focusableItems: false,
};

export const ListContextProvider = listContext.Provider;
export const useListContext = () => React.useContext(listContext) ?? listContextDefaultValue;
