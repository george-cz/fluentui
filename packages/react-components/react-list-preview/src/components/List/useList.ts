import * as React from 'react';
import { getIntrinsicElementProps, slot, useEventCallback } from '@fluentui/react-utilities';
import { useArrowNavigationGroup } from '@fluentui/react-tabster';
import { ListLayout, ListProps, ListState, IList } from './List.types';
import { useListFeatures } from '../../hooks/useListFeatures';
import { useListSelection } from '../../hooks/useListSelection';

const useComponentRef = (componentRef: React.Ref<IList> | undefined, selection: ListState['selection']) => {
  React.useImperativeHandle(
    componentRef,
    () => ({
      selection,
    }),
    [selection],
  );
};

/**
 * Create the state required to render List.
 *
 * The returned state can be modified with hooks such as useListStyles_unstable,
 * before being passed to renderList_unstable.
 *
 * @param props - props from this instance of List
 * @param ref - reference to root HTMLElement of List
 */
export const useList_unstable = (props: ListProps, ref: React.Ref<HTMLElement>): ListState => {
  const {
    layout = ListLayout.Vertical,
    focusableItems = false,
    customArrowNavigationOptions,
    selectable = false,
    selectionMode = 'multiselect',
    defaultSelectedItems,
    componentRef,
    onSelectionChange,
  } = props;

  const arrowNavigationAttributes = useArrowNavigationGroup({
    axis: layout === ListLayout.Grid ? 'grid-linear' : 'both',
    memorizeCurrent: true,
    ...(customArrowNavigationOptions || {}),
  });

  const [items, setItems] = React.useState<Array<{ id: string | number }>>([]);

  const { selection } = useListFeatures({ items }, [
    useListSelection({ defaultSelectedItems, onSelectionChange, selectionMode }),
  ]);
  useComponentRef(componentRef, selection);

  const registerItem = useEventCallback((id: string | number) => {
    if (!items.find(item => item.id === id)) {
      setItems(current => [...current, { id }]);
    }
  });

  const deregisterItem = useEventCallback((id: string | number) => {
    if (items.find(k => k.id === id)) {
      setItems(current => current.filter(item => item.id !== id));
    }
  });

  return {
    components: {
      root: 'ul',
    },
    root: slot.always(
      getIntrinsicElementProps('ul', {
        ref,
        tabIndex: -1,
        ...selection.getListProps(),
        ...arrowNavigationAttributes,
        ...props,
      }),
      { elementType: 'ul' },
    ),
    layout,
    // context:
    focusableItems,
    items,
    registerItem: selectable ? registerItem : undefined,
    deregisterItem: selectable ? deregisterItem : undefined,
    // only pass down selection state if its handled internally, otherwise just report the events
    selection: selectable ? selection : undefined,
  };
};
