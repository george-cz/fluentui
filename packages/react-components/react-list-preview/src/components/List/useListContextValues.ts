import { ListContextValues, ListState } from './List.types';

export function useListContextValues_unstable(state: ListState): ListContextValues {
  const { selection, as, navigationMode, listItemRole, validateListItems } = state;

  const listContext = {
    selection,
    as,
    listItemRole,
    navigationMode,
    validateListItems,
  };

  return {
    listContext,
  };
}
