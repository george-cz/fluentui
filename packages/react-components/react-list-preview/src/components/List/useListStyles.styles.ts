import type { SlotClassNames } from '@fluentui/react-utilities';
import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { ListSlots, ListState } from './List.types';

export const listClassNames: SlotClassNames<ListSlots> = {
  root: 'fui-List',
  // TODO: add class names for all slots on ListSlots.
  // Should be of the form `<slotName>: 'fui-List__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    ...shorthands.padding(0),
    ...shorthands.margin(0),
    textIndent: 0,
    listStyleType: 'none',
  },

  rootHorizontal: {
    display: 'flex',
  },

  rootGrid: {
    display: 'grid',
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the List slots based on the state
 */
export const useListStyles_unstable = (state: ListState): ListState => {
  const styles = useStyles();

  const layoutToStyles = {
    ['vertical']: styles.root,
    ['horizontal']: styles.rootHorizontal,
    ['grid']: styles.rootGrid,
  };

  state.root.className = mergeClasses(
    listClassNames.root,
    styles.root,
    layoutToStyles[state.layout],
    state.root.className,
  );

  return state;
};
