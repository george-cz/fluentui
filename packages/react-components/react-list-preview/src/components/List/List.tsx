import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useList_unstable } from './useList';
import { renderList_unstable } from './renderList';
import { useListStyles_unstable } from './useListStyles.styles';
import type { ListProps } from './List.types';
import { useListContextValues_unstable } from './useListContextValues';

/**
 * List component - TODO: add more docs
 */
export const List: ForwardRefComponent<ListProps> = React.forwardRef((props, ref) => {
  const state = useList_unstable(props, ref);
  const listContext = useListContextValues_unstable(state);

  useListStyles_unstable(state);
  // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
  useCustomStyleHook_unstable('useListStyles_unstable')(state);

  return renderList_unstable(state, listContext);
});

List.displayName = 'List';
