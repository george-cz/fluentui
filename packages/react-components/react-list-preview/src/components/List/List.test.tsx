import '@testing-library/jest-dom';

import * as React from 'react';
import { fireEvent, prettyDOM, render, waitFor, within } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { List } from './List';
import { ListProps } from './List.types';
import { ListItem } from '../ListItem/ListItem';
import { EventHandler } from 'react';
import userEvent from '@testing-library/user-event';

function expectListboxItemSelected(item: HTMLElement, selected: boolean) {
  expect(item.getAttribute('aria-selected')).toBe(selected.toString());
}

describe('List', () => {
  isConformant({
    Component: List as React.FunctionComponent<ListProps>,
    displayName: 'List',
    testOptions: {
      'consistent-callback-args': {
        // onSelectionChange has an eventArgument which is React.SyntheticEvent. This throws an error during testing
        ignoreProps: ['onSelectionChange'],
      },
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(
      <List>
        <ListItem value="test-value-1">First ListItem</ListItem>
        <ListItem value="test-value-2">Second ListItem</ListItem>
      </List>,
    );
    expect(result.container).toMatchSnapshot();
  });

  describe('roles', () => {
    it('default - should have list/listitem roles', () => {
      const result = render(
        <List>
          <ListItem>First ListItem</ListItem>
          <ListItem>Second ListItem</ListItem>
        </List>,
      );

      expect(result.getAllByRole('list')).toHaveLength(1);
      expect(result.getAllByRole('listitem')).toHaveLength(2);
    });

    it('selectable - should have listbox/option roles', () => {
      const result = render(
        <List selectable>
          <ListItem>First ListItem</ListItem>
          <ListItem>Second ListItem</ListItem>
        </List>,
      );

      expect(result.getAllByRole('listbox')).toHaveLength(1);
      expect(result.getAllByRole('option')).toHaveLength(2);
    });

    // TODO: Add more tests for multiple action once those components are created
  });

  describe('selection', () => {
    it('Single mode should unselect previous', () => {
      const result = render(
        <List selectable selectionMode="single">
          <ListItem value="test-value-1">First ListItem</ListItem>
          <ListItem value="test-value-2">Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      const secondItem = result.getByText('Second ListItem');

      expectListboxItemSelected(firstItem, false);
      expectListboxItemSelected(secondItem, false);

      firstItem.click();
      expectListboxItemSelected(firstItem, true);
      expectListboxItemSelected(secondItem, false);

      secondItem.click();
      expectListboxItemSelected(firstItem, false);
      expectListboxItemSelected(secondItem, true);
    });

    it('Multi mode should select an item when clicked', () => {
      const result = render(
        <List selectable selectionMode="multiselect">
          <ListItem value="test-value-1">First ListItem</ListItem>
          <ListItem value="test-value-2">Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      const secondItem = result.getByText('Second ListItem');

      // [ ][ ]
      expectListboxItemSelected(firstItem, false);
      expectListboxItemSelected(secondItem, false);

      firstItem.click();
      // [x][ ]
      expectListboxItemSelected(firstItem, true);
      expectListboxItemSelected(secondItem, false);

      secondItem.click();
      // [x][x]
      expectListboxItemSelected(firstItem, true);
      expectListboxItemSelected(secondItem, true);

      secondItem.click();
      // [x][ ]
      expectListboxItemSelected(firstItem, true);
      expectListboxItemSelected(secondItem, false);

      firstItem.click();
      // [ ][ ]
      expectListboxItemSelected(firstItem, false);
      expectListboxItemSelected(secondItem, false);
    });
  });

  describe('mouse behavior', () => {
    describe('no selection', () => {
      it('Click should trigger onClick', () => {
        const onClick = jest.fn();

        const result = render(
          <List>
            <ListItem onClick={onClick}>First ListItem</ListItem>
            <ListItem>Second ListItem</ListItem>
          </List>,
        );

        const firstItem = result.getByText('First ListItem');
        firstItem.click();
        expect(onClick).toHaveBeenCalledTimes(1);
      });
    });

    describe('with selection', () => {
      function interactWithFirstElement(
        interaction: (firstItem: HTMLElement) => void,
        customOnclick?: EventHandler<React.SyntheticEvent<HTMLElement>>,
      ) {
        const onClick = jest.fn(customOnclick);

        const result = render(
          <List selectable selectionMode="multiselect">
            <ListItem onClick={onClick}>First ListItem</ListItem>
            <ListItem>Second ListItem</ListItem>
          </List>,
        );

        const firstItem = result.getByText('First ListItem');
        interaction(firstItem);

        return { listItem: firstItem, onClick };
      }

      it('Click should trigger selection and callback by default', () => {
        const { listItem, onClick } = interactWithFirstElement(item => item.click());
        expect(onClick).toHaveBeenCalledTimes(1);
        expectListboxItemSelected(listItem, true);
      });

      it('preventDefault should prevent selection', () => {
        const { listItem, onClick } = interactWithFirstElement(
          item => item.click(),
          e => e.preventDefault(),
        );
        expect(onClick).toHaveBeenCalledTimes(1);
        expectListboxItemSelected(listItem, false);
      });

      it("Click on the checkbox should trigger selection, onClick shouldn't be called", () => {
        const { listItem, onClick } = interactWithFirstElement(item => {
          within(item).getByRole('checkbox').click();
        });
        expect(onClick).not.toHaveBeenCalled();
        expectListboxItemSelected(listItem, true);
      });
    });
  });

  describe('focusable list items', () => {
    it('should not be focusable by default', () => {
      const result = render(
        <List>
          <ListItem>First ListItem</ListItem>
          <ListItem>Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      firstItem.focus();
      expect(document.activeElement).not.toBe(firstItem);
    });
    it('should be focusable when onClick is passed', () => {
      const result = render(
        <List>
          <ListItem onClick={() => null}>First ListItem</ListItem>
          <ListItem onClick={() => null}>Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      firstItem.focus();
      expect(document.activeElement).toBe(firstItem);
    });
    it('should be focusable when onKeyDown is passed', () => {
      const result = render(
        <List>
          <ListItem onKeyDown={() => null}>First ListItem</ListItem>
          <ListItem onKeyDown={() => null}>Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      firstItem.focus();
      expect(document.activeElement).toBe(firstItem);
    });
    it('should be focusable when list is selectable', () => {
      const result = render(
        <List selectable>
          <ListItem>First ListItem</ListItem>
          <ListItem>Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      firstItem.focus();
      expect(document.activeElement).toBe(firstItem);
    });
    it('should be focusable when tabIndex=0 is passed', () => {
      const result = render(
        <List selectable>
          <ListItem tabIndex={0}>First ListItem</ListItem>
          <ListItem tabIndex={0}>Second ListItem</ListItem>
        </List>,
      );

      const firstItem = result.getByText('First ListItem');
      firstItem.focus();
      expect(document.activeElement).toBe(firstItem);
    });
  });

  describe('keyboard behavior', () => {
    describe('no selection', () => {
      function pressKeyOnListItem(key: string) {
        const onClick = jest.fn();

        const result = render(
          <List>
            <ListItem onClick={onClick}>First ListItem</ListItem>
            <ListItem>Second ListItem</ListItem>
          </List>,
        );

        const firstItem = result.getByText('First ListItem');
        fireEvent.keyDown(firstItem, { key });
        return { onClick };
      }

      it('should NOT trigger onClick when random key is pressed', () => {
        expect(pressKeyOnListItem('a').onClick).toHaveBeenCalledTimes(0);
      });
      it('Space should trigger onClick', () => {
        expect(pressKeyOnListItem(' ').onClick).toHaveBeenCalledTimes(1);
      });
      it('Enter should trigger onClick', () => {
        expect(pressKeyOnListItem('Enter').onClick).toHaveBeenCalledTimes(1);
      });
    });

    describe('with selection', () => {
      function pressOnListItem(key: string, customOnclick?: EventHandler<React.SyntheticEvent<HTMLElement>>) {
        const onClick = jest.fn(customOnclick);

        const result = render(
          <List selectable selectionMode="multiselect">
            <ListItem onClick={onClick}>First ListItem</ListItem>
            <ListItem>Second ListItem</ListItem>
          </List>,
        );

        const firstItem = result.getByText('First ListItem');
        fireEvent.keyDown(firstItem, { key });
        return { onClick, listItem: firstItem };
      }

      it('Spacebar toggles selection by default, onClick is not called', () => {
        const { onClick, listItem } = pressOnListItem(' ');
        expect(onClick).not.toHaveBeenCalled();
        expectListboxItemSelected(listItem, true);
      });

      it('Enter toggles selection by default, onClick is called', () => {
        const { onClick, listItem } = pressOnListItem('Enter');
        expect(onClick).toHaveBeenCalledTimes(1);
        expectListboxItemSelected(listItem, true);
      });
      it("Enter doesn't toggle selection if default is prevented", () => {
        const { onClick, listItem } = pressOnListItem('Enter', e => e.preventDefault());
        expect(onClick).toHaveBeenCalledTimes(1);
        expectListboxItemSelected(listItem, false);
      });
    });
  });
});
