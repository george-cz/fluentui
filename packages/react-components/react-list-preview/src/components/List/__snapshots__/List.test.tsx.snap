// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List rendering render as div and div doesn't throw 1`] = `
<div>
  <div
    class="fui-List"
    data-tabster="{\\"mover\\":{\\"cyclic\\":false,\\"direction\\":1,\\"memorizeCurrent\\":true}}"
  >
    <div
      class="fui-ListItem"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2},\\"focusable\\":{\\"ignoreKeydown\\":{\\"Enter\\":true}}}"
      data-value="test-value-1"
      id="test-value-1"
      role="listitem"
    >
      First ListItem
    </div>
    <div
      class="fui-ListItem"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2},\\"focusable\\":{\\"ignoreKeydown\\":{\\"Enter\\":true}}}"
      data-value="test-value-2"
      id="test-value-2"
      role="listitem"
    >
      Second ListItem
    </div>
  </div>
</div>
`;

exports[`List rendering renders a default state 1`] = `
<div>
  <ul
    class="fui-List"
    data-tabster="{\\"mover\\":{\\"cyclic\\":false,\\"direction\\":1,\\"memorizeCurrent\\":true}}"
  >
    <li
      class="fui-ListItem"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2},\\"focusable\\":{\\"ignoreKeydown\\":{\\"Enter\\":true}}}"
      data-value="test-value-1"
      id="test-value-1"
      role="listitem"
      value="test-value-1"
    >
      First ListItem
    </li>
    <li
      class="fui-ListItem"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2},\\"focusable\\":{\\"ignoreKeydown\\":{\\"Enter\\":true}}}"
      data-value="test-value-2"
      id="test-value-2"
      role="listitem"
      value="test-value-2"
    >
      Second ListItem
    </li>
  </ul>
</div>
`;
